charger/pulsares.go

```diff
@@ -0,0 +1,158 @@
+)
+
+// Pulsares charger implementation
+type Pulsares struct {
+	conn *modbus.Connection
+	curr uint16
+}
+
+const (
+	pulsaresRegStatus  = 0x1f
+	pulsaresRegCurrent = 0x5d
+)
+
+func init() {
+	registry.Add("Pulsares", NewPulsaresFromConfig)
+}
+
+// NewPulsaresFromConfig creates a Pulsares charger from generic config
+func NewPulsaresFromConfig(other map[string]interface{}) (api.Charger, error) {
+	cc := struct {
+		modbus.Settings `mapstructure:",squash"`
+	}{
+		Settings: modbus.Settings{
+			ID: 1,
+		},
+	}
+
+	if err := util.DecodeOther(other, &cc); err != nil {
+		return nil, err
+	}
+
+	return NewPulsares(cc.URI, cc.Device, cc.Comset, cc.Baudrate, modbus.ProtocolFromRTU(cc.RTU), cc.ID)
+}
+
+// NewPulsares creates Pulsares charger
+func NewPulsares(uri, device, comset string, baudrate int, proto modbus.Protocol, slaveID uint8) (api.Charger, error) {
+	conn, err := modbus.NewConnection(uri, device, comset, baudrate, proto, slaveID)
+	if err != nil {
+		return nil, err
+	}
+
+	log := util.NewLogger("pulsares")
+	conn.Logger(log.TRACE)
+
+	wb := &Pulsares{
+		conn: conn,
+		curr: 6000,
+	}
+
+	// get initial state from charger
+	curr, err := wb.getCurrent()
+	if err != nil {
+		return nil, fmt.Errorf("current limit: %w", err)
+	}
+	if curr >= 6000 {
+		wb.curr = curr
+	}
+
+	return wb, err
+}
+
+func (wb *Pulsares) setCurrent(current uint16) error {
+	b := make([]byte, 2)
+	binary.BigEndian.PutUint16(b, uint16(current))
+
+	_, err := wb.conn.WriteMultipleRegisters(pulsaresRegCurrent, 1, b)
+
+	return err
+}
+
+func (wb *Pulsares) getCurrent() (uint16, error) {
+	b, err := wb.conn.ReadHoldingRegisters(pulsaresRegCurrent, 1)
+	if err != nil {
+		return 0, err
+	}
+
+	return binary.BigEndian.Uint16(b), nil
+}
+
+// Status implements the api.Charger interface
+func (wb *Pulsares) Status() (api.ChargeStatus, error) {
+	b, err := wb.conn.ReadHoldingRegisters(pulsaresRegStatus, 1)
+	if err != nil {
+		return api.StatusNone, err
+	}
+
+	switch u := binary.BigEndian.Uint16(b); u {
+	case 0:
+		return api.StatusA, nil
+	case 1, 2:
+		return api.StatusB, nil
+	case 3, 4:
+		return api.StatusC, nil
+	default:
+		return api.StatusNone, fmt.Errorf("invalid status: %d", u)
+	}
+}
+
+// Enabled implements the api.Charger interface
+func (wb *Pulsares) Enabled() (bool, error) {
+	curr, err := wb.getCurrent()
+
+	return curr >= 6000, err
+}
+
+// Enable implements the api.Charger interface
+func (wb *Pulsares) Enable(enable bool) error {
+	var curr uint16
+	if enable {
+		curr = wb.curr
+	}
+
+	return wb.setCurrent(curr)
+}
+
+// MaxCurrent implements the api.Charger interface
+func (wb *Pulsares) MaxCurrent(current int64) error {
+	return wb.MaxCurrentMillis(float64(current))
+}
+
+var _ api.ChargerEx = (*Pulsares)(nil)
+
+// MaxCurrent implements the api.ChargerEx interface
+func (wb *Pulsares) MaxCurrentMillis(current float64) error {
+	if current < 6 {
+		return fmt.Errorf("invalid current %.1f", current)
+	}
+
+	wb.curr = uint16(current * 1e3)
```

Comment:
gem. anderen Boxen sollte wb.curr immer *erst* gesetzt werden wenn das Schreiben erfolgreich war!

Reply:
So hatte ich das auch zuerst und hab es dann von einer anderen WB so kopiert weil es kompakter war und es auch schon ziemlich spät war.
Aber genau das habe ich mir auch gedacht.

Reply:
Noch etwas, wenn ich den bisherigen Stand richtig verstehe, wird die serielle RS485 verwendet. Das ist zwar soweit OK aber erfordert zusätzlichen Verkabelungsaufwand. Ich hatte damals (mit Manu7788 ?) darüber gesprochen und ein EW11, WIFI zu RS485 Adapter, bekommen, den ich dann im Gehäuse eingeklebt habe. Die Spannungsversorgung dafür kommt aus der Box. Ich denke, auch andere potentielle User wollen nicht noch mehr Kabel ziehen. Könnt Ihr das bitte berücksichtigen.

Reply:
@premultiply weisst du boch welche andere? Sollten wir da auch korrigieren.

Reply:
> Könnt Ihr das bitte berücksichtigen.

Wird es automatisch

Reply:
> Noch etwas, wenn ich den bisherigen Stand richtig verstehe, wird die serielle RS485 verwendet. Das ist zwar soweit OK aber erfordert zusätzlichen Verkabelungsaufwand. Ich hatte damals (mit Manu7788 ?) darüber gesprochen und ein EW11, WIFI zu RS485 Adapter, bekommen, den ich dann im Gehäuse eingeklebt habe. Die Spannungsversorgung dafür kommt aus der Box. Ich denke, auch andere potentielle User wollen nicht noch mehr Kabel ziehen. Könnt Ihr das bitte berücksichtigen.

Genau, wir hatten gesprochen. Das ist ein Modbus-TCP auf RTU-Adapter. 

Bezüglich des neuen Raspberry Pi Managers: Hier werden die seriellen Schnittstellen benutzt, die dem System bekannt sind. Wichtig ist, dass das RTS Signal der Schnittstelle beim Senden auf high gesetzt wird. Ich habe das in c so getestet:

`
rs485conf.flags |= SER_RS485_RTS_ON_SEND;
rs485conf.flags &= ~(SER_RS485_RTS_AFTER_SEND);

rs485conf.delay_rts_before_send = 1;
rs485conf.delay_rts_after_send = 1;
`

Ich weiß jetzt nicht wie und ob das in der modbus.go bereits so umgesetzt ist. Für den Manager wird das benötigt um den Transceiver umzuschalten.

Reply:
> @premultiply weisst du boch welche andere? Sollten wir da auch korrigieren.

daheimladen-mb

Reply:
Ist behoben

Reply:
Gerne Mail an info@evcc.io

Reply:
> > Noch etwas, wenn ich den bisherigen Stand richtig verstehe, wird die serielle RS485 verwendet. Das ist zwar soweit OK aber erfordert zusätzlichen Verkabelungsaufwand. Ich hatte damals (mit Manu7788 ?) darüber gesprochen und ein EW11, WIFI zu RS485 Adapter, bekommen, den ich dann im Gehäuse eingeklebt habe. Die Spannungsversorgung dafür kommt aus der Box. Ich denke, auch andere potentielle User wollen nicht noch mehr Kabel ziehen. Könnt Ihr das bitte berücksichtigen.
> 
> Genau, wir hatten gesprochen. Das ist ein Modbus-TCP auf RTU-Adapter.
> 
> Bezüglich des neuen Raspberry Pi Managers: Hier werden die seriellen Schnittstellen benutzt, die dem System bekannt sind. Wichtig ist, dass das RTS Signal der Schnittstelle beim Senden auf high gesetzt wird. Ich habe das in c so getestet:
> 
> ` rs485conf.flags |= SER_RS485_RTS_ON_SEND; rs485conf.flags &= ~(SER_RS485_RTS_AFTER_SEND);
> 
> rs485conf.delay_rts_before_send = 1; rs485conf.delay_rts_after_send = 1; `
> 
> Ich weiß jetzt nicht wie und ob das in der modbus.go bereits so umgesetzt ist. Für den Manager wird das benötigt um den Transceiver umzuschalten.

Okay, das Problem hat sich grade erledigt. Das kann ich auf dem System so lösen, dass nur per Serial-Interface geschrieben werden muss, so wie ihr es aktuell macht. 

Compilieren von EVCC auf dem PI geht auch, hatte aber RAM Probleme bei 1 GByte. Warte auf den neuen PI mit 8 GByte, dann kann ich den Branch hier aufbauen und testen.

Reply:
Go kann cross-compilieren:

    GOOS=linux GOARCH=arm make

Reply:
> Go kann cross-compilieren:
> 
> ```
> GOOS=linux GOARCH=arm make
> ```

Hat geklappt und startet. Allerdings:

`cannot create charger 'wallbox1': cannot create charger 'template': template not found: Pulsares`

`git clone --single-branch --branch device/pulsares https://github.com/evcc-io/evcc.git`

Die pulsares.go ist da und die json hat den Eintrag.

Reply:
Dann stimmt noch irgendwas mit der Konfig oder dem Template nicht.
Kannst auch mit `type: pulsares` testen.

Reply:
> Dann stimmt noch irgendwas mit der Konfig oder dem Template nicht. Kannst auch mit `type: pulsares` testen.

geht auch nicht.

`# open evcc at http://evcc.local:7070
network:
  schema: http
  host: evcc.local # .local suffix announces the hostname on MDNS
  port: 7070

log: debug
levels:
  cache: error

# unique installation id
plant: a5d6c5959a0aab4b0096ea46ef8aaa4379a4b3522dbddd84e11aa35c2b8196e4

interval: 10s # control cycle interval

sponsortoken: eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJleHAiOjE2OTcyOTU2MDAsImlhdCI6MTY5NjQzMTYwMCwiaXNzIjoiZXZjYy5pbyIsInN1YiI6IlRyaWFsIFNwb25zb3JzaGlwIPCfkKUiLCJzcGUiOnRydWUsInNyYyI6Im1hIn0.bRmIqYKLKAjfP1u4rfKxfHqWr-tm5sUhruq3wawWAwQ

# sponsors can set telemetry: true to enable anonymous data aggregation
# see https://github.com/evcc-io/evcc/discussions/4554
telemetry: false

chargers:
- type: template
  template: pulsares 
  id: 2  
  device: pulcharge  
  baudrate: 57600  
  comset: 8N1  
  modbus: /dev/ttyAMA1  
  name: wallbox1

loadpoints:
- title: Tisch
  charger: wallbox1
  mode: now
  phases: 3
  mincurrent: 6
  maxcurrent: 32
  resetOnDisconnect: true

site:
  title: Home
  meters:
`

Reply:
> Dann stimmt noch irgendwas mit der Konfig oder dem Template nicht. Kannst auch mit `type: pulsares` testen.

[config.txt](https://github.com/evcc-io/evcc/files/12821630/config.txt)

Reply:
abl hat funktioniert, ich sehe eine ausgabe am logic analyzer.